ethers-middleware/src/builder.rs

```diff
@@ -0,0 +1,67 @@
+use ethers_providers::Middleware;
+
+/// A builder struct useful to compose different [`Middleware`](ethers_providers::Middleware) layers
+/// and then build a composed [`Provider`](ethers_providers::Provider) architecture.
+/// [`Middleware`](ethers_providers::Middleware) composition acts in a wrapping fashion. Adding a
+/// new layer results in wrapping its predecessor.
+///
+/// Builder can be used as follows:
+///
+/// ```rust
+/// use ethers_providers::{Middleware, Provider, Http};
+/// use std::sync::Arc;
+/// use std::convert::TryFrom;
+/// use ethers_signers::{LocalWallet, Signer};
+/// use ethers_middleware::{*,gas_escalator::*,gas_oracle::*};
+///
+/// fn example() {
+///     let key = "fdb33e2105f08abe41a8ee3b758726a31abdd57b7a443f470f23efce853af169";
+///     let signer = key.parse::<LocalWallet>().unwrap();
+///     let address = signer.address();
+///     let escalator = GeometricGasPrice::new(1.125, 60_u64, None::<u64>);
+///
+///     let provider = Provider::<Http>::try_from("http://localhost:8545").unwrap();
+///
+///     ProviderBuilder::from(provider)
+///         .wrap_into(|p| GasEscalatorMiddleware::new(p, escalator, Frequency::PerBlock))
+///         .wrap_into(|p| SignerMiddleware::new(p, signer))
+///         .wrap_into(|p| GasOracleMiddleware::new(p, EthGasStation::new(None)))
+///         .wrap_into(|p| NonceManagerMiddleware::new(p, address)) // Outermost layer
+///         .build();
+/// }
+/// ```
+pub struct ProviderBuilder<M> {
+    inner: Option<M>,
```

Comment:
why is this an option if all uses of it immediately `take()` and `unwrap()`?

Reply:
I found it handy to take the ownership on the inner Middleware and pass it through. Maybe restricting the M bound to Clone could work as well. Other suggestions?

Reply:
you can take ownership of the `inner` by changing `wrap_into` and build to consume `self` as follows. 

```
pub fn wrap_into<F, R>(self, builder_fn: F) -> ProviderBuilder<R>
    where
        F: FnOnce(M) -> R,
        R: Middleware,
    {
        let provider = self.inner;
        let provider = builder_fn(provider.unwrap());
        ProviderBuilder::from(provider)
    }

    /// Returns the overall[`Middleware`](ethers_providers::Middleware) as a reference to the
    /// outermost layer
    pub fn build(self) -> M {
        self.inner
    }
}```


Consuming is best-practice for a builder anyway, because it prevents misuse. Current design permits bad usage in sensible-looking code like this:

```
// normal looking code
let mut builder = ProviderBuilder::from(provider);

// A user who hasn't read the docs may not know that chained access is required
// this `take()` s from builder and returns a provider builder that the user doesn't bind
builder.wrap_into(|p| SomeMiddleware::new(p);

// this panics because the `wrap_into` took the `inner` and builder is now empty
builder.build()
```